### PR TITLE
Simplify Base58 encoding/decoding

### PR DESCRIPTION
**Pull request by zvonand**
Get rid of Base-X library in contrib -- it is too slow. Replaced by a simple C-like code snippet.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improved Base58 encoding/decoding.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Let's not add a data type domain.

Reason: data type domains are bug-prone, and it's not the best architectural thing in ClickHouse.
We want to rewrite the data type domains to something else.
And base58 is not going to be really popular.

**Comment by zvonand** [comment]
> Let's not add a data type domain.
Yes, I understand that the domain will not be widely adopted. But, for example, for storing and later analyzing cryptocurrencies data, saving 17% of storage may be visible

**Comment by zvonand** [comment]
Can I have `can-be-tested` here?

**Comment by alexey-milovidov** [comment]
@zvonand Let's not introduce a new data type domain because otherwise, you will fall into a trap of introducing Base64 and Hex domains, that look more important.

**Comment by alexey-milovidov** [comment]
We don't want new data type domains before https://github.com/ClickHouse/ClickHouse/issues/18568
and before fixing the tech debt after introduction of data type domains: https://github.com/ClickHouse/ClickHouse/issues/14567

File: src/Functions/FunctionBase58Conversion.h [hunk 1]
```diff
@@ -26,72 +24,39 @@ struct Base58Encode
...
-                dst_data.resize(current_allocated_size);
-                auto processed_offset = dst_pos - dst;
-                dst = dst_data.data();
-                dst_pos = dst;
-                dst_pos += processed_offset;
-            }
-            std::memcpy(dst_pos, encoded.c_str(), ++outlen);
+            encodeBase58(source, dst_pos);
+
+            size_t encoded_length = strlen(reinterpret_cast<const char *>(dst_pos));
```
Comment (rschu1ze) [inline src/Functions/FunctionBase58Conversion.h:53]: Hm, could the encoded size not be calculated on-the-fly in `encodeBase58()`?

File: src/Common/base58.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        dst[c_idx] = base58_encoding_alphabet[static_cast<unsigned char>(dst[c_idx])];
+    }
+    dst[idx] = '\0';
+    return true;
+}
+
+inline bool decodeBase58(const char8_t * src, char8_t * dst)
+{
+    const char map_digits[128]
```
[the remark this answers is not in the document]
Comment (zvonand) [reply src/Common/base58.h:46]: Well, I sticked to C style on purpose

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
```
Jul 21 14:44:37 In file included from ../src/Functions/FunctionsBase58.cpp:1:
Jul 21 14:44:37 In file included from ../src/Functions/FunctionBase58Conversion.h:9:
Jul 21 14:44:37 /build/src/Common/base58.h:47:12: error: constant expression evaluates to -1 which cannot be narrowed to type 'char' [-Wc++11-narrowing]
Jul 21 14:44:37         = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
Jul 21 14:44:37            ^~
...
Jul 21 14:44:37 fatal error: too many errors emitted, stopping now [-ferror-limit=]
Jul 21 14:44:37 20 errors generated.
```

Happens only on ARM and PPC because `char` is signed on these and unsigned on x86. I found the least ugly solution something like this:

```
const unsigned char um = UINT_MAX;                                                                                                     
const unsigned char map_digits[128]
  = {um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, 
     um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, um, 0,  1,  2,  3,  4,  5,  6,  7,  8,  um, um, um, um,   um, um, 
     um, 9,  10, 11, 12, 13, 14, 15, 16, um, 17, 18, 19, 20, 21, um, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, um, um, um, um, um, 
     um, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, um, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, um, um, um, um, um};
```

As you can see, it retains the alignment and it neither needs explicit casts a la ```(unsigned char)(-1)``` (char types also have no literal suffix to their avail) nor needs to disable the warning.

**Comment by alexey-milovidov** [comment]
> because char is signed on these and unsigned on x86

The other way around.

**Comment by alexey-milovidov** [comment]
@rschu1ze Cultural issues were found: #41185

Examples:
- :warning: **no comments in code** :vomiting_face:;
- the function is marked inline and put into the header file for no reason;
- usage of architecture-dependent types (unsigned int, signed char);
- multiple effects in a single statement (like `x[*ptr++] = *src++`);
- cryptic conversions between int and char;
- usage of C-style macro UINT_MAX;
- using `for` statement where `while` is more appropriate;

**Comment by alexey-milovidov** [comment]
There are even more flaws in the code:
- there is no `tryBase58Decode` function;
- it is using `auto` when it is unneeded;
- the library functions expect zero-terminated string with no reason *and no comments*;

**Comment by zvonand** [comment]
@alexey-milovidov 
> there is no tryBase58Decode function;

Why would we need it? If smth is wrong, exception is thrown


> cryptic conversions between int and char;
> multiple effects in a single statement (like x[*ptr++] = *src++)

This code is (obviously) C-styled, and thus IMO it is fine. Maybe I'm just too damaged with C experience 🥲 


> it is using auto when it is unneeded

Not the only place in CH 😃  As a relatively new person here, I can say this is awful, esp. when you try to understand data types transformations and conversions, but I thought it was some way of making code more readable
OK, I will reduce number of further used `auto`s

**Comment by alexey-milovidov** [comment]
Thank you!

The `tryBase58Decode` function is needed because we have the `tryBase64Decode` function.

**Comment by alexey-milovidov** [comment]
We still need to think on how to optimize the algorithm.

**Pull request merged**